### PR TITLE
Refactor of bx to be closer to the lua build scripts

### PR DESCRIPTION
**Pull request by bwrsandman**
Mirror the build script directory of bx.
Closely follow the lua script definitions and options. Converted them to cmake equivalents.
Add bin2c and lemon to tools.
- [x] Add tests

Still need to test the cmake function for
- [x] bin2c

### CODE REVIEW

**Review by bkaradzic — COMMENTED** [review]
You should actually ignore lemon.

**Review by bkaradzic — APPROVED** [review]

File: cmake/bgfxToolUtils.cmake [hunk 1]
```diff
@@ -1,3 +1,74 @@
...
+  )
+  bin2c_parse(CLI INPUT_FILE ${ARGS_INPUT_FILE} OUTPUT_FILE ${ARGS_OUTPUT_FILE} ARRAY_NAME ${ARGS_ARRAY_NAME})
+  add_custom_command(OUTPUT ${ARGS_OUTPUT_FILE} COMMAND bgfx::bin2c ${CLI})
+endfunction()
+
+# add_lemon_command(
+#	FILE filename
+#)
+#
+function (add_bin2c_command)
```
Comment (bwrsandman) [inline cmake/bgfxToolUtils.cmake]:
```suggestion
function (add_lemon_command)
```

Comment (bkaradzic) [reply cmake/bgfxToolUtils.cmake]: It looks like you removed bin2c and added lemon?! Remove lemon. :)

**Review by exdal — APPROVED** [review]

File: cmake/bgfxToolUtils.cmake (hunk 1 above)
Comment (bwrsandman) [reply cmake/bgfxToolUtils.cmake]: This is a copy paste mistake on my part when adding a function for lemon.
It'll be gone by removing lemon.

**Comment by bwrsandman** [comment]
@e-erdal @bkaradzic could you take a look at the addition to the README and let me know what you think.

File: README.md [hunk 2]
```diff
@@ -20,19 +18,105 @@ cmake --build cmake-build
...
+```
+This defines a shaderc command to generate headers for a number of `TYPE` shaders with `SHADERS` files and `VARYING_DEF` file in the `OUTPUT_DIR` directory. There will be one generated shader for each supported rendering API on this current platform according to the `BGFX_EMBEDDED_SHADER` macro in `bgfx/embedded_shader.h`.
+
+The generated headers will have names in the format of `${SHADERS}.${RENDERING_API}.bin.h` where `RENDERING_API` can be `glsl`, `essl`, `spv`, `dx9`, `dx11` and `mtl` depending on the availability of the platform.
+
+Adding these `SHADERS` as source files to a target will run `shaderc` at build time and they will rebuild if either the contents of the `SHADERS` or the `VARYING_DEF` change.
+
+#### Examples: Generating shaders as headers
+```cmake
+compile_shader_to_header(
```
Comment (bwrsandman) [inline README.md]:
```suggestion
bgfx_compile_shader_to_header(
```

File: README.md [hunk 3]
```diff
@@ -20,19 +18,105 @@ cmake --build cmake-build
...
+  TYPE VERTEX
+  SHADERS vs.sc
+  VARYING_DEF varying.def.sc
+  OUTPUT_DIR ${CMAKE_BINARY_DIR}/include/generated/shaders
+)
+compile_shader_to_header(
```
Comment (bwrsandman) [inline README.md]:
```suggestion
bgfx_compile_shader_to_header(
```

**Review by bkaradzic — APPROVED** [review]

**Pull request merged**